src/main/java/com/streamr/client/StreamrClient.java

```diff
@@ -291,16 +306,44 @@ public Subscription subscribe(Stream stream, int partition, MessageHandler handl
     public void unsubscribe(Subscription sub) {
         UnsubscribeRequest unsubscribeRequest = new UnsubscribeRequest(sub.getStreamId(), sub.getPartition());
         sub.setState(Subscription.State.UNSUBSCRIBING);
+        sub.setResending(false);
         this.websocket.send(unsubscribeRequest.toJson());
     }
 
     private void handleSubcribeResponse(SubscribeResponse res) throws SubscriptionNotFoundException {
         Subscription sub = subs.get(res.getStreamId(), res.getStreamPartition());
         sub.setState(Subscription.State.SUBSCRIBED);
+        if (sub.hasResendOptions()) {
+            ResendOption resendOption = sub.getEffectiveResendOption();
+            this.websocket.send(resendOption.toRequest(res.getStreamId(), res.getStreamPartition(), sub.getId(), this.getSessionToken()).toJson());
+        }
     }
 
     private void handleUnsubcribeResponse(UnsubscribeResponse res) throws SubscriptionNotFoundException {
         Subscription sub = subs.get(res.getStreamId(), res.getStreamPartition());
         sub.setState(Subscription.State.UNSUBSCRIBED);
     }
+
+    private void handleResendResponseResending(ResendResponseResending res) throws SubscriptionNotFoundException {
+        Subscription sub = subs.get(res.getStreamId(), res.getStreamPartition());
+        log.debug("Resending started for subscription "+sub.getId());
+    }
+
+    private void handleResendResponseNoResend(ResendResponseNoResend res) throws SubscriptionNotFoundException {
+        Subscription sub = subs.get(res.getStreamId(), res.getStreamPartition());
+        endResendAndCheckQueue(sub);
+    }
+
+    private void handleResendResponseResent(ResendResponseResent res) throws SubscriptionNotFoundException {
+        Subscription sub = subs.get(res.getStreamId(), res.getStreamPartition());
+        endResendAndCheckQueue(sub);
+    }
+
+    private void endResendAndCheckQueue(Subscription sub) {
```

Comment:
IMO might be nicer to have something like:

- call `sub.startResend()` instead of explicitly setting `sub.setResending(true)`
- call `sub.endResend()` instead of explicitly setting `sub.setResending(false)` and calling `handleQueue` from outside